File: java/src/main/java/com/github/apstndb/spanvalue/TypeFormat.java
package com.github.apstndb.spanvalue;

import static com.github.apstndb.spanvalue.ProtoAccess.arrayElementType;
import static com.github.apstndb.spanvalue.ProtoAccess.fieldName;
import static com.github.apstndb.spanvalue.ProtoAccess.fieldType;
import static com.github.apstndb.spanvalue.ProtoAccess.protoTypeFqn;
import static com.github.apstndb.spanvalue.ProtoAccess.structFields;
import static com.github.apstndb.spanvalue.ProtoAccess.typeAnnotation;
import static com.github.apstndb.spanvalue.ProtoAccess.typeCode;

import java.util.ArrayList;
import java.util.List;

/** Format Cloud Spanner google.spanner.v1.Type values. */
public final class TypeFormat {
  private TypeFormat() {}

  public enum StructMode {
    BASE,
    RECURSIVE,
    RECURSIVE_WITH_NAME
  }

  public enum ProtoEnumMode {
    BASE,
    LEAF,
    FULL,
    LEAF_WITH_KIND,
    FULL_WITH_KIND
  }

  public enum ArrayMode {
    BASE,
    RECURSIVE
  }

  public enum UnknownMode {
    UNKNOWN,
    TYPE_CODE,
    VERBOSE,
    PANIC
  }

  public enum TypeAnnotationMode {
    SUFFIX,
    OMIT,
    PRIMARY
  }

  public record FormatOption(
      StructMode structMode,
      ProtoEnumMode proto,
      ProtoEnumMode enumMode,
      ArrayMode array,
      UnknownMode unknown,
      TypeAnnotationMode typeAnnotation) {

    public FormatOption() {
      this(
          StructMode.BASE,
          ProtoEnumMode.BASE,
          ProtoEnumMode.BASE,
          ArrayMode.BASE,
          UnknownMode.UNKNOWN,
          TypeAnnotationMode.SUFFIX);
    }
  }

  public static final FormatOption FORMAT_OPTION_SIMPLEST =
      new FormatOption(
          StructMode.BASE,
          ProtoEnumMode.BASE,
          ProtoEnumMode.BASE,
          ArrayMode.BASE,
          UnknownMode.TYPE_CODE,
          TypeAnnotationMode.SUFFIX);

  public static final FormatOption FORMAT_OPTION_SIMPLE =
      new FormatOption(
          StructMode.BASE,
          ProtoEnumMode.LEAF,
          ProtoEnumMode.LEAF,
          ArrayMode.RECURSIVE,
          UnknownMode.UNKNOWN,
          TypeAnnotationMode.SUFFIX);

  public static final FormatOption FORMAT_OPTION_NORMAL =
      new FormatOption(
          StructMode.RECURSIVE,
          ProtoEnumMode.LEAF,
          ProtoEnumMode.LEAF,
          ArrayMode.RECURSIVE,
          UnknownMode.VERBOSE,
          TypeAnnotationMode.SUFFIX);

  public static final FormatOption FORMAT_OPTION_VERBOSE =
      new FormatOption(
          StructMode.RECURSIVE_WITH_NAME,
          ProtoEnumMode.FULL,
          ProtoEnumMode.FULL,
          ArrayMode.RECURSIVE,
          UnknownMode.VERBOSE,
          TypeAnnotationMode.SUFFIX);

  public static final FormatOption FORMAT_OPTION_MORE_VERBOSE =
      new FormatOption(
          StructMode.RECURSIVE_WITH_NAME,
          ProtoEnumMode.FULL_WITH_KIND,
          ProtoEnumMode.FULL_WITH_KIND,
          ArrayMode.RECURSIVE,
          UnknownMode.VERBOSE,
          TypeAnnotationMode.SUFFIX);

  private static String lastCut(String s, String sep) {
    int idx = s.lastIndexOf(sep);
    return idx >= 0 ? s.substring(idx + 1) : s;
  }

  private static String annotationSuffix(int ann) {
    if (ann == TypeAnnotationCode.TYPE_ANNOTATION_CODE_UNSPECIFIED.getValue()) {
      return "";
    }
    String name = TypeAnnotationCode.nameFor(ann);
    if (name == null) {
      return "(" + ann + ")";
    }
    return "(" + name + ")";
  }

  private static String annotationName(int ann) {
    String name = TypeAnnotationCode.nameFor(ann);
    return name != null ? name : String.valueOf(ann);
  }

  public static String formatTypeCode(int code, UnknownMode mode) {
    String name = TypeCode.nameFor(code);
    if (name != null) {
      return name;
    }
    return switch (mode) {
      case TYPE_CODE -> String.valueOf(code);
      case VERBOSE -> "UNKNOWN(" + code + ")";
      case PANIC -> throw new UnknownTypeException("unknown TypeCode(" + code + ")");
      default -> "UNKNOWN";
    };
  }

  public static String formatProtoEnum(Object typ, ProtoEnumMode mode) {
    int code = typeCode(typ);
    String fqn = protoTypeFqn(typ);
    String codeName = formatTypeCode(code, UnknownMode.VERBOSE);
    return switch (mode) {
      case LEAF -> lastCut(fqn, ".");
      case FULL -> fqn;
      case LEAF_WITH_KIND -> codeName + "<" + lastCut(fqn, ".") + ">";
      case FULL_WITH_KIND -> codeName + "<" + fqn + ">";
      default -> codeName;
    };
  }

  public static String formatStructFields(List<Object> fields, FormatOption option) {
    List<String> parts = new ArrayList<>();
    for (Object field : fields) {
      String typeStr = formatType(fieldType(field), option);
      if (option.structMode() == StructMode.RECURSIVE_WITH_NAME && !fieldName(field).isEmpty()) {
        parts.add(fieldName(field) + " " + typeStr);
      } else {
        parts.add(typeStr);
      }
    }
    return String.join(", ", parts);
  }

  private static String formatTypeImpl(Object typ, FormatOption option) {
    int code = typeCode(typ);
    if (code == TypeCode.ARRAY.getValue() && option.array() != ArrayMode.BASE) {
      return "ARRAY<" + formatType(arrayElementType(typ), option) + ">";
    }
    if (code == TypeCode.PROTO.getValue()) {
      return formatProtoEnum(typ, option.proto());
    }
    if (code == TypeCode.ENUM.getValue()) {
      return formatProtoEnum(typ, option.enumMode());
    }
    if (code == TypeCode.STRUCT.getValue() && option.structMode() != StructMode.BASE) {
      return "STRUCT<" + formatStructFields(structFields(typ), option) + ">";
    }
    return formatTypeCode(code, option.unknown());
  }

  public static String formatType(Object typ, FormatOption option) {
    if (option == null) {
      option = FORMAT_OPTION_SIMPLE;
    }
    int ann = typeAnnotation(typ);
    return switch (option.typeAnnotation()) {
      case OMIT -> formatTypeImpl(typ, option);
      case PRIMARY -> {
        if (ann != TypeAnnotationCode.TYPE_ANNOTATION_CODE_UNSPECIFIED.getValue()) {
          yield annotationName(ann);
        }
        yield formatTypeImpl(typ, option);
      }
      case SUFFIX -> formatTypeImpl(typ, option) + annotationSuffix(ann);
    };
  }

  public static String formatTypeSimplest(Object typ) {
    return formatType(typ, FORMAT_OPTION_SIMPLEST);
  }

  public static String formatTypeSimple(Object typ) {
    return formatType(typ, FORMAT_OPTION_SIMPLE);
  }

  public static String formatTypeNormal(Object typ) {
    return formatType(typ, FORMAT_OPTION_NORMAL);
  }

  public static String formatTypeVerbose(Object typ) {
    return formatType(typ, FORMAT_OPTION_VERBOSE);
  }

  public static String formatTypeMoreVerbose(Object typ) {
    return formatType(typ, FORMAT_OPTION_MORE_VERBOSE);
  }

  public static String formatTypeVerboseAnnotationOmit(Object typ) {
    return formatType(
        typ,
        new FormatOption(
            FORMAT_OPTION_VERBOSE.structMode(),
            FORMAT_OPTION_VERBOSE.proto(),
            FORMAT_OPTION_VERBOSE.enumMode(),
            FORMAT_OPTION_VERBOSE.array(),
            FORMAT_OPTION_VERBOSE.unknown(),
            TypeAnnotationMode.OMIT));
  }

  public static String formatTypeVerboseAnnotationPrimary(Object typ) {
    return formatType(
        typ,
        new FormatOption(
            FORMAT_OPTION_VERBOSE.structMode(),
            FORMAT_OPTION_VERBOSE.proto(),
            FORMAT_OPTION_VERBOSE.enumMode(),
            FORMAT_OPTION_VERBOSE.array(),
            FORMAT_OPTION_VERBOSE.unknown(),
            TypeAnnotationMode.PRIMARY));
  }
}

File: csharp/src/Apstndb.SpanValue/TypeFormat.cs
namespace Apstndb.SpanValue;

public enum StructMode
{
    Base = 0,
    Recursive = 1,
    RecursiveWithName = 2,
}

public enum ProtoEnumMode
{
    Base = 0,
    Leaf = 1,
    Full = 2,
    LeafWithKind = 3,
    FullWithKind = 4,
}

public enum ArrayMode
{
    Base = 0,
    Recursive = 1,
}

public enum UnknownMode
{
    Unknown = 0,
    TypeCode = 1,
    Verbose = 2,
    Panic = 3,
}

public enum TypeAnnotationMode
{
    Suffix = 0,
    Omit = 1,
    Primary = 2,
}

public sealed record FormatOption(
    StructMode Struct = StructMode.Base,
    ProtoEnumMode Proto = ProtoEnumMode.Base,
    ProtoEnumMode Enum = ProtoEnumMode.Base,
    ArrayMode Array = ArrayMode.Base,
    UnknownMode Unknown = UnknownMode.Unknown,
    TypeAnnotationMode TypeAnnotation = TypeAnnotationMode.Suffix)
{
    public FormatOption WithTypeAnnotation(TypeAnnotationMode mode) =>
        this with { TypeAnnotation = mode };
}

public static class TypeFormat
{
    public static readonly FormatOption FormatOptionSimplest = new(
        Unknown: UnknownMode.TypeCode);

    public static readonly FormatOption FormatOptionSimple = new(
        Proto: ProtoEnumMode.Leaf,
        Enum: ProtoEnumMode.Leaf,
        Array: ArrayMode.Recursive);

    public static readonly FormatOption FormatOptionNormal = new(
        Struct: StructMode.Recursive,
        Proto: ProtoEnumMode.Leaf,
        Enum: ProtoEnumMode.Leaf,
        Array: ArrayMode.Recursive,
        Unknown: UnknownMode.Verbose);

    public static readonly FormatOption FormatOptionVerbose = new(
        Struct: StructMode.RecursiveWithName,
        Proto: ProtoEnumMode.Full,
        Enum: ProtoEnumMode.Full,
        Array: ArrayMode.Recursive,
        Unknown: UnknownMode.Verbose);

    public static readonly FormatOption FormatOptionMoreVerbose = new(
        Struct: StructMode.RecursiveWithName,
        Proto: ProtoEnumMode.FullWithKind,
        Enum: ProtoEnumMode.FullWithKind,
        Array: ArrayMode.Recursive,
        Unknown: UnknownMode.Verbose);

    public static string FormatTypeCode(int code, UnknownMode mode = UnknownMode.Verbose)
    {
        var name = Codes.TypeCodeName(code);
        if (name is not null)
            return name;

        return mode switch
        {
            UnknownMode.TypeCode => code.ToString(),
            UnknownMode.Verbose => $"UNKNOWN({code})",
            UnknownMode.Panic => throw new UnknownTypeException($"unknown TypeCode({code})"),
            _ => "UNKNOWN",
        };
    }

    public static string FormatProtoEnum(object? typ, ProtoEnumMode mode)
    {
        var code = Proto.TypeCode(typ);
        var fqn = Proto.ProtoTypeFqn(typ);
        var codeName = FormatTypeCode(code);

        return mode switch
        {
            ProtoEnumMode.Leaf => LastCut(fqn, "."),
            ProtoEnumMode.Full => fqn,
            ProtoEnumMode.LeafWithKind => $"{codeName}<{LastCut(fqn, ".")}>",
            ProtoEnumMode.FullWithKind => $"{codeName}<{fqn}>",
            _ => codeName,
        };
    }

    public static string FormatStructFields(IReadOnlyList<object?> fields, FormatOption option)
    {
        var parts = new List<string>(fields.Count);
        foreach (var field in fields)
        {
            var typeStr = FormatType(Proto.FieldType(field), option);
            var name = Proto.FieldName(field);
            if (option.Struct == StructMode.RecursiveWithName && !string.IsNullOrEmpty(name))
                parts.Add($"{name} {typeStr}");
            else
                parts.Add(typeStr);
        }

        return string.Join(", ", parts);
    }

    public static string FormatType(object? typ, FormatOption? option = null)
    {
        option ??= FormatOptionSimple;

        var ann = Proto.TypeAnnotation(typ);
        if (option.TypeAnnotation == TypeAnnotationMode.Omit)
            return FormatTypeImpl(typ, option);

        if (option.TypeAnnotation == TypeAnnotationMode.Primary)
        {
            if (ann != (int)TypeAnnotationCode.TypeAnnotationCodeUnspecified)
                return AnnotationName(ann);
            return FormatTypeImpl(typ, option);
        }

        return FormatTypeImpl(typ, option) + AnnotationSuffix(ann);
    }

    public static string FormatTypeSimplest(object? typ) => FormatType(typ, FormatOptionSimplest);
    public static string FormatTypeSimple(object? typ) => FormatType(typ, FormatOptionSimple);
    public static string FormatTypeNormal(object? typ) => FormatType(typ, FormatOptionNormal);
    public static string FormatTypeVerbose(object? typ) => FormatType(typ, FormatOptionVerbose);
    public static string FormatTypeMoreVerbose(object? typ) => FormatType(typ, FormatOptionMoreVerbose);

    public static string FormatTypeVerboseAnnotationOmit(object? typ) =>
        FormatType(typ, FormatOptionVerbose.WithTypeAnnotation(TypeAnnotationMode.Omit));

    public static string FormatTypeVerboseAnnotationPrimary(object? typ) =>
        FormatType(typ, FormatOptionVerbose.WithTypeAnnotation(TypeAnnotationMode.Primary));

    private static string FormatTypeImpl(object? typ, FormatOption option)
    {
        var code = Proto.TypeCode(typ);

        if (code == (int)TypeCode.Array && option.Array != ArrayMode.Base)
        {
            var elem = Proto.ArrayElementType(typ);
            return $"ARRAY<{FormatType(elem, option)}>";
        }

        if (code == (int)TypeCode.Proto)
            return FormatProtoEnum(typ, option.Proto);

        if (code == (int)TypeCode.Enum)
            return FormatProtoEnum(typ, option.Enum);

        if (code == (int)TypeCode.Struct && option.Struct != StructMode.Base)
            return $"STRUCT<{FormatStructFields(Proto.StructFields(typ), option)}>";

        return FormatTypeCode(code, option.Unknown);
    }

    private static string LastCut(string s, string sep)
    {
        var idx = s.LastIndexOf(sep, StringComparison.Ordinal);
        return idx >= 0 ? s[(idx + sep.Length)..] : s;
    }

    private static string AnnotationSuffix(int ann)
    {
        if (ann == (int)TypeAnnotationCode.TypeAnnotationCodeUnspecified)
            return "";

        var name = Codes.TypeAnnotationName(ann);
        return name is null ? $"({ann})" : $"({name})";
    }

    private static string AnnotationName(int ann)
    {
        var name = Codes.TypeAnnotationName(ann);
        return name ?? ann.ToString();
    }
}
